controllers/nodemodulesconfig_reconciler.go

```diff
@@ -0,0 +1,681 @@
+		Named(NodeModulesConfigReconcilerName).
+		For(&kmmv1beta1.NodeModulesConfig{}).
+		Owns(
+			&v1.Pod{},
+			builder.WithPredicates(filter.HasLabel(actionLabelKey)),
+		).
+		// TODO maybe replace this with Owns() if we make nodes the owners of NodeModulesConfigs.
+		Watches(
+			&v1.Node{},
+			handler.EnqueueRequestsFromMapFunc(nodeToNMCMapFunc),
+			builder.WithPredicates(filter.SkipDeletions()),
+		).
+		Complete(r)
+}
+
+func workerPodName(nodeName, moduleName string) string {
+	return fmt.Sprintf("kmm-worker-%s-%s", nodeName, moduleName)
+}
+
+func GetContainerStatus(statuses []v1.ContainerStatus, name string) v1.ContainerStatus {
+	for i := range statuses {
+		if statuses[i].Name == name {
+			return statuses[i]
+		}
+	}
+
+	return v1.ContainerStatus{}
+}
+
+func FindNodeCondition(cond []v1.NodeCondition, conditionType v1.NodeConditionType) *v1.NodeCondition {
+	for i := 0; i < len(cond); i++ {
+		c := cond[i]
+
+		if c.Type == conditionType {
+			return &c
+		}
+	}
+
+	return nil
+}
+
+//go:generate mockgen -source=nodemodulesconfig_reconciler.go -package=controllers -destination=mock_nodemodulesconfig_reconciler.go WorkerHelper
+
+type WorkerHelper interface {
+	ProcessModuleSpec(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig, spec *kmmv1beta1.NodeModuleSpec, status *kmmv1beta1.NodeModuleStatus) error
+	ProcessOrphanModuleStatus(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig, status *kmmv1beta1.NodeModuleStatus) error
+	SyncStatus(ctx context.Context, nmc *kmmv1beta1.NodeModulesConfig) error
+	RemoveOrphanFinalizers(ctx context.Context, nodeName string) error
+}
+
+type workerHelper struct {
+	client client.Client
+	pm     PodManager
+}
+
+func NewWorkerHelper(client client.Client, pm PodManager) WorkerHelper {
+	return &workerHelper{
+		client: client,
+		pm:     pm,
+	}
+}
+
+func (w *workerHelper) ProcessModuleSpec(
+	ctx context.Context,
+	nmc *kmmv1beta1.NodeModulesConfig,
+	spec *kmmv1beta1.NodeModuleSpec,
+	status *kmmv1beta1.NodeModuleStatus,
+) error {
+	logger := ctrl.LoggerFrom(ctx)
+
+	statusNil := false
+
+	if status == nil {
+		statusNil = true
+	} else if status.InProgress {
+		logger.Info("Sync status is in progress; skipping")
+		return nil
+	}
+
+	if statusNil || status.Config == nil {
+		logger.Info("Missing status or config; creating loader Pod")
+
+		return w.pm.CreateLoaderPod(ctx, nmc, spec)
+	}
+
+	if !spec.Config.Equal(status.Config) {
+		logger.Info("Outdated config in status; creating unloader Pod")
+
+		return w.pm.CreateUnloaderPod(ctx, nmc, status)
+	}
+
+	node := v1.Node{}
+
+	if err := w.client.Get(ctx, types.NamespacedName{Name: nmc.Name}, &node); err != nil {
+		return fmt.Errorf("could not get node %s: %v", nmc.Name, err)
+	}
+
+	readyCondition := FindNodeCondition(node.Status.Conditions, v1.NodeReady)
+	if readyCondition == nil {
+		return fmt.Errorf("node %s has no Ready condition", nmc.Name)
+	}
+
+	if status.LastTransitionTime.Before(&readyCondition.LastTransitionTime) {
```

Comment:
A comment describing the scenarion (why we need to compare times) would be beneficial here

Reply:
Added a comment.

Reply:
if you can add that the reason to compare the times is in order to catch the node reboot scenario, it would be great

Reply:
Improved the comment. Does it look OK to you?